### PR TITLE
[SYCL] Add missing builtins for vec<X, 1>

### PR DESCRIPTION
**Pull request by alexbatashev**
On host using builtins with `vec<X, 1>` types causes linkage errors. Add missing overloads.

### CODE REVIEW

File: sycl/source/detail/builtins_helper.hpp [hunk 1]
```diff
@@ -153,44 +161,51 @@
...
   __MAKE_SR_1V_OR(Fun, Call, 4, Ret, Arg1)                                     \
   __MAKE_SR_1V_OR(Fun, Call, 8, Ret, Arg1)                                     \
   __MAKE_SR_1V_OR(Fun, Call, 16, Ret, Arg1)
 
 #define MAKE_1V_2P(Fun, Ret, Arg1, Arg2)                                       \
+  __MAKE_1V_2P(Fun, 1, Ret, Arg1, Arg2)                                        \
   __MAKE_1V_2P(Fun, 2, Ret, Arg1, Arg2)                                        \
   __MAKE_1V_2P(Fun, 3, Ret, Arg1, Arg2)                                        \
   __MAKE_1V_2P(Fun, 4, Ret, Arg1, Arg2)                                        \
   __MAKE_1V_2P(Fun, 8, Ret, Arg1, Arg2) __MAKE_1V_2P(Fun, 16, Ret, Arg1, Arg2)
```
Comment (v-klochkov) [inline sycl/source/detail/builtins_helper.hpp:199]: Minor comment: it is not about your change, but shouldn't this line be split into 2 lines. Same for Line 168.

File: sycl/source/detail/builtins_helper.hpp [hunk 2]
```diff
@@ -70,40 +73,42 @@
...
+    sycl::vec<Ret, N> r;                                                       \
     detail::helper<N - 1>().run_1v_2p(                                         \
         r, [](Arg1 x, Arg2 *y) { return cl::__host_std::Fun(x, y); }, x, y);   \
     return r;                                                                  \
   }
 
 #define __MAKE_1V_2V_3P(Fun, N, Ret, Arg1, Arg2, Arg3)                         \
-  __SYCL_EXPORT Ret##N Fun __NOEXC(Arg1##N x, Arg2##N y, Arg3##N *z) {         \
-    Ret##N r;                                                                  \
+  __SYCL_EXPORT sycl::vec<Ret, N> Fun __NOEXC(                                 \
```
Comment (v-klochkov) [inline sycl/source/detail/builtins_helper.hpp:109]: It is for remquo intrinsic. Why do you think it is __NOEXC ?
std::remquo may raise FE_INVALID in some cases (https://en.cppreference.com/w/cpp/numeric/math/remquo), or is it somewhat different?

File: sycl/source/detail/builtins_helper.hpp [hunk 3]
```diff
@@ -153,44 +161,51 @@
...
+  __MAKE_1V_2V_RS(Fun, Call, 1, Ret, Arg1, Arg2)                               \
   __MAKE_1V_2V_RS(Fun, Call, 2, Ret, Arg1, Arg2)                               \
   __MAKE_1V_2V_RS(Fun, Call, 3, Ret, Arg1, Arg2)                               \
   __MAKE_1V_2V_RS(Fun, Call, 4, Ret, Arg1, Arg2)
 
 #define MAKE_1V_2V_3P(Fun, Ret, Arg1, Arg2, Arg3)                              \
+  __MAKE_1V_2V_3P(Fun, 1, Ret, Arg1, Arg2, Arg3)                               \
   __MAKE_1V_2V_3P(Fun, 2, Ret, Arg1, Arg2, Arg3)                               \
   __MAKE_1V_2V_3P(Fun, 3, Ret, Arg1, Arg2, Arg3)                               \
   __MAKE_1V_2V_3P(Fun, 4, Ret, Arg1, Arg2, Arg3)                               \
```
Comment (v-klochkov) [inline sycl/source/detail/builtins_helper.hpp:211]: Shouldn't 8 and 16 element vectors be in this list as well?

**Comment by v-klochkov** [comment]
Please add a trivial LIT test or extend some existing test.

File: sycl/source/detail/builtins_helper.hpp (hunk 3 above)
Comment (v-klochkov) [reply sycl/source/detail/builtins_helper.hpp:211]: + same for other types of intrinsics mentioned above. e.g. lines: 205, 135,146, ...

File: sycl/source/detail/builtins_helper.hpp (hunk 2 above)
Comment (v-klochkov) [reply sycl/source/detail/builtins_helper.hpp:109]: Sorry, you did not change the attributes here. My mistake. I am removing the comment.

**Comment by s-kanaev** [comment]
@alexbatashev ping

File: sycl/source/detail/builtins_helper.hpp (hunk 1 above)
Comment (alexbatashev) [reply sycl/source/detail/builtins_helper.hpp:199]: I'd prefer to address this separately.

File: sycl/source/detail/builtins_helper.hpp (hunk 3 above)
Comment (alexbatashev) [reply sycl/source/detail/builtins_helper.hpp:211]: Also added as part of PR

**Comment by alexbatashev** [comment]
> Please add a trivial LIT test or extend some existing test.

Our current tests are not really scalable, and covering >500 builtins in one PR looks to much for me. Is it ok if I add a simple smoke test for now, and do a followup with better testing?

**Comment by bader** [comment]
> Is it ok if I add a simple smoke test for now, and do a followup with better testing?

I think adding a simple smoke test is enough for LIT testing, we are not aiming to cover 100% of use cases with LIT tests.
Better testing can be provided by SYCL-CTS and/or separate test suite.

**Comment by alexbatashev** [comment]
> > Is it ok if I add a simple smoke test for now, and do a followup with better testing?
> 
> I think adding a simple smoke test is enough for LIT testing, we are not aiming to cover 100% of use cases with LIT tests.
> Better testing can be provided by SYCL-CTS and/or separate test suite.

Then I’ll look into why CTS didn’t catch missing builtins.

**Comment by v-klochkov** [comment]
> Then I’ll look into why CTS didn’t catch missing builtins.

Adding LIT test with few (3-5) intrinsics would be enough, it would show that the patch works and to have a basic check to avoid regressions in future.

File: sycl/test/regression/builtins_vector1.cpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
+// RUN: %clangxx -fsycl %s -o %t.out
```
Comment (v-klochkov) [inline sycl/test/regression/builtins_vector1.cpp:1]:
```suggestion
// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
```

**Comment by alexbatashev** [comment]
@v-klochkov I found even more missing builtins, and I believe some are still missing. I'll do a followup on adding the rest of them once proper testing is in place.

**Pull request merged**